pkg/notify/notify.go

```diff
@@ -173,75 +225,115 @@ func toIssueLabel(groupLabels alertmanager.KV) string {
 	query := fmt.Sprintf("project=\"%s\" and labels=%q order by resolutiondate desc", project, issueLabel)
 	options := &jira.SearchOptions{
 		Fields:     []string{"summary", "status", "resolution", "resolutiondate"},
 		MaxResults: 2,
 	}
-	level.Debug(logger).Log("msg", "search", "query", query, "options", options)
-	issues, resp, err := r.client.Issue.Search(query, options)
+
+	level.Debug(r.logger).Log("msg", "search", "query", query, "options", fmt.Sprintf("%+v", options))
+	issues, resp, err := r.client.Search(query, options)
 	if err != nil {
-		retry, err := handleJiraError("Issue.Search", resp, err, logger)
+		retry, err := handleJiraErrResponse("Issue.Search", resp, err, r.logger)
 		return nil, retry, err
 	}
-	if len(issues) > 0 {
-		if len(issues) > 1 {
-			// Swallow it, but log a message.
-			level.Debug(logger).Log("msg", "  more than one issue matched, picking most recently resolved", "query", query, "issues", issues)
-		}
 
-		level.Debug(logger).Log("msg", "  found", "issue", issues[0], "query", query)
-		return &issues[0], false, nil
+	if len(issues) == 0 {
+		level.Debug(r.logger).Log("msg", "no results", "query", query)
+		return nil, false, nil
 	}
-	level.Debug(logger).Log("msg", "  no results", "query", query)
-	return nil, false, nil
+
+	issue := issues[0]
+	if len(issues) > 1 {
+		level.Warn(r.logger).Log("msg", "more than one issue matched, picking most recently resolved", "query", query, "issues", issues, "picked", issue)
+	}
+
+	level.Debug(r.logger).Log("msg", "found", "issue", issue, "query", query)
+	return &issue, false, nil
 }
 
-func (r *Receiver) reopen(issueKey string, logger log.Logger) (bool, error) {
-	transitions, resp, err := r.client.Issue.GetTransitions(issueKey)
+func (r *Receiver) issueToReuse(project string, issueGroupLabel string) (*jira.Issue, bool, error) {
+	issue, retry, err := r.search(project, issueGroupLabel)
 	if err != nil {
-		return handleJiraError("Issue.GetTransitions", resp, err, logger)
+		return nil, retry, err
+	}
+
+	if issue == nil {
+		return nil, false, nil
+	}
+
+	resolutionTime := time.Time(issue.Fields.Resolutiondate)
+	if resolutionTime != (time.Time{}) && resolutionTime.Add(time.Duration(*r.conf.ReopenDuration)).Before(r.timeNow()) {
+		level.Debug(r.logger).Log("msg", "potential issue to reuse was resolved after reopen duration, skipping", "key", issue.Key, "label", issueGroupLabel, "resolution_time", resolutionTime.Format(time.RFC3339), "reopen_duration", *r.conf.ReopenDuration)
+		return nil, false, nil
 	}
+	// Reuse issue.
+	return issue, false, nil
+}
+
+func (r *Receiver) updateSummary(issueKey string, summary string) (bool, error) {
+	level.Debug(r.logger).Log("msg", "updating issue with new summary", "key", issueKey, "summary", summary)
+
+	issueUpdate := &jira.Issue{
+		Key: issueKey,
+		Fields: &jira.IssueFields{
+			Summary: summary,
+		},
+	}
+	issue, resp, err := r.client.UpdateWithOptions(issueUpdate, nil)
```

Comment:
Yes, it would. But I just found it more cleaner to use explicit method. IMO API that has thousands of ways of doing exactly the same thing (e.g just to help user to not write 4 chars `, nil`) is a really antipattern, an `go-jira` has it everywhere. Let's use explicit thing in our code at least. What do you think?

Reply:
I'm usually in favor of whatever is easier on the eyes/brain. If `go-jira` was our code, then I'd consider this argument and likely agree with it. Even though if I had to guess I'd say `Update` was there first so it's still there for backward compatibility.

But I'm not a religious warrior (in this respect, I clearly am in others) so as long as both options are there I'd go with the one that reads more easily.

That being said, if this truly bothers you, you can leave it as is, I guess. Although the `go-jira` maintainer(s) will likely never find out about your objections to their API design. (o:

Reply:
> That being said, if this truly bothers you, you can leave it as is, I guess. Although the go-jira maintainer(s) will likely never find out about your objections to their API design. (o:

That's not my intention (: But to change the world to be better place worth to start from our own code ;p Not a bit decision here, I would leave this although I guess it's not YAGNI (we don't plan to use opts in near future)